### ISSUE TITLE
LRPs in k8s Spike  Option 3 from previous Spike

### ISSUE THREAD

**Issue by andrew-edgar** [julz/cube#34]
See previous spike and try out option 3 from story #30

**Comment by mnitchev** [comment]
Update on the spike 

## TL; DR
Looks like the changes should be fairly straightforward for the purposes of the spike. We have started making them in the CC to try it out. We need an environment where we can deploy CC and Eirini so we can test.

# Challenges

## LRPs in K8s: 

To complete the story k8s/list.go needs to be implemented together with an lrp handler. For `start` endpoint Desirer should be used.
Listing kube deployments is pretty simple. The following is all we need (excluding mapping kube Deployment to LRP):
```golang
client.AppsV1beta1().Deployments(namespace).List(av1.ListOptions{})
```

## Converting CC Process Model to Eirini LRP Model

 *  on the Eirini side we need:
```golang
        Name            string
        Image           string
        Command         []string
        Env             map[string]string
        TargetInstances int
```
For example the eirini handler could exepct the following json in its endpoint:
```json
{
    "name": "appname_guid",
    "image": "url?",
    "command": ["ls", "-l", "-s"],
    "env": {
        "GOPATH": "/go",
        "PORT": "8080"
    },
    "targetInstances": 4
}
```
*  in CC we have the [ProcessModel](https://github.com/cloudfoundry/cloud_controller_ng/blob/master/app/models/runtime/process_model.rb) object that describes LRPs. It seems like it has the following attributes, but since these come from the databse, this should be investigated further: 
```ruby
:name, :production, :space_guid, :stack_guid, :buildpack,
:detected_buildpack, :detected_buildpack_guid,:environment_json, :memory, :instances, :disk_quota,
:state, :version, :command, :console, :debug,:staging_task_id,
:package_state, :health_check_type, :health_check_timeout,:health_check_http_endpoint,
:staging_failed_reason, :staging_failed_description, :diego,:docker_image, :package_updated_at,
:detected_start_command, :enable_ssh, :ports
```
In our case  `:name`, `:docker_image`, `:command`, `:environment_json`, `:instances` (or the method `desired_instances`) seem like all the things we need.

## Implementing the OPI Client

* `AppRecipeBuilder` invocations in [ProcessesSync](https://github.com/cloudfoundry/cloud_controller_ng/blob/master/lib/cloud_controller/diego/processes_sync.rb) should be moved in `Diego::Client` or in BBS [Stager](https://github.com/cloudfoundry/cloud_controller_ng/blob/5d5ea91a958d6bc07b74fae935c21178f3dc28eb/lib/cloud_controller/diego/bbs_stager_client.rb)/[Apps](https://github.com/cloudfoundry/cloud_controller_ng/blob/5d5ea91a958d6bc07b74fae935c21178f3dc28eb/lib/cloud_controller/diego/bbs_apps_client.rb)/[Task](https://github.com/cloudfoundry/cloud_controller_ng/blob/5d5ea91a958d6bc07b74fae935c21178f3dc28eb/lib/cloud_controller/diego/bbs_task_client.rb) clients. 

*  `OPI::ApsClient` replaces [BbsAppsClient](https://github.com/cloudfoundry/cloud_controller_ng/blob/5d5ea91a958d6bc07b74fae935c21178f3dc28eb/lib/cloud_controller/diego/bbs_apps_client.rb). This seems to be the simplest way to make the change. Our client would have to implement the following methods:
```ruby
def desire_app(lrp)
def fetch_scheduling_infos
def update_app(process_guid, lrp_update) #stub
def get_app(process_guid) #stub
def stop_app(process_guid) #stub
```
Since `BbsAppsClient` has more methods we should investigate if they should also be stubbed.

* The new client needs to be instantiated in the [DependencyLocator](https://github.com/cloudfoundry/cloud_controller_ng/blob/9882f75af7cd01b82465e88a2742fef834acfc56/lib/cloud_controller/dependency_locator.rb#L391)

* To try this approach for the spike, we are going to need a pipeline with a bosh lite director and so we can deploy CF with the modified Cloud Controller.
